### DISCUSSION TITLE
`MiniRocketMultivariate` - what is the algorithm in the estimator?

### DISCUSSION THREAD

**Discussion by 4d30** [alan-turing-institute/sktime#3298] [Q&A]
I have multiple timeseries for an instance, but they do not occur simultaneously.

Section 1.2 of examples/AA_datatypes_and_datasets.ipynb indicates that for each **variable**, a point in time is indicated by the index of **time points**.

>     Section 1.2: Time series panels - the "Panel" scitype
>     In the "numpy3D" mtype, time series panels are represented by an in-memory container obj: np.ndarray as follows.
> 
>     structure convention: obj must be 3D, i.e., obj.shape must have length 2.
>     instances: instances correspond to axis 0 elements of obj.
>     instance index: the instance index is implicit and by-convention. The i-th element of axis 0 (for an integer i) is interpreted as indicative of observing instance i.
>     variables: variables correspond to axis 1 elements of obj.
>     variable names: the "numpy3D" mtype cannot represent variable names.
>     time points: time points correspond to axis 2 elements of obj.
>     time index: the time index is implicit and by-convention. The i-th elemtn of axis 2 (for an integer i) is interpreted as an observation at the time point i.
>     capabilities: can represent panels of multivariate series; cannot represent unequally spaced series; cannot represent panels of unequally supported series; cannot represent panels of series with different sets of variables.

When fitting MiniRocketMultvariate with this scitype I am curious of how the actual convolutions are performed over the multiple variables?

Are the transformations performed on each **variable** independently? (Multiple 1-D transforms)

Or is the transformation performed on all **variables** combined? (1 Multi-D transform)

**Comment by fkiraly** [comment]
Hm, the docstring does not seem to say. The quoted reference is about the univariate method.

According to the commit log, the author is @fstinner.

@fstinner, could you kindly explain?

Formally, the estimator is lacking a sufficient scientific reference.
Would you be able to add to the docstring what is happening? Or, at least, leave an explanation here, so someone else can add it in?

FYI both, @4d30, @fstinner, transformers vectorize automatically over variables now, so the univariate `MultiRocket` is applied on a per-variable basis (multiple 1D transforms).

Does the multivariate version do the same, or sth different?

**Comment by fkiraly** [comment]
> I have multiple timeseries for an instance, but they do not occur simultaneously.

PS, if that is the case, then `numpy3D` is not the type you should use, since it can only represent time series where the all instances and variables are observed at the same time indices.

Assuming we are using the terms "instances" and "time points" in the same way.

**Comment by fstinner** [comment]
I am not the original author of MiniRocketMultivariate but had only adapted the docstrings of the Rockets. But the problem is the same for MultiRocket. The indices of multivariate and univariate (long list) are the same as far as I know, so the multivariate version should extract the features for each time series of a data source independently. Otherwise, the indices should be different according to my understanding.

I hope to get to it this week, but there's a tight schedule.

**Comment by 4d30** [comment]
> PS, if that is the case, then numpy3D is not the type you should use, since it can only represent time series where the all instances and variables are observed at the same time indices.

@fkiraly  This seems like a strange requirement to me. **That is, is it a requirement that numpy3d[0,0,0] occur contemporaneously with numpy3d[1,0,0] AND numpy3d[0,1,0]?**

I made some figures to facilitate what it is that I am asking about. If these are multiple 1d transforms over a  **variable** number of  dimensions, then information from one **variable** ought to not bleed into another. Then something like this might work:

**[Idea 1]**
![rocket_fig](https://user-images.githubusercontent.com/45021583/185964263-01c90472-5b81-4605-9cfa-2799a83ab753.svg)

On the other hand, multiple transforms could be performed for each colored "event" in this figure. The transform output could be concatenated all of the transforms for the events (for a total of ~30,000 features), and classifier fit atop of that.

**[Idea 2]**
![rocket_fig2](https://user-images.githubusercontent.com/45021583/185967647-fd837c9b-15eb-44fd-abe8-473bd4500b5a.svg)

The way I am interpreting the docs and this discussion is suggests that implementing **[idea 2]** would  be the path forward.

**Reply by fkiraly** [reply]
>That is, is it a requirement that numpy3d[0,0,0] occur contemporaneously with numpy3d[1,0,0] AND numpy3d[0,1,0]?

It is an assumption in the data type, yes - although it could be contemporaneous in relative time, not necessarily in absolute time-
E.g., all could be "5 min after start of the experiment", where "start of the experiment" may be different in absolute time.

The issue is less with 3D numpy array than with the assumptions of how the algorithms interpret it.
